proposals/debug-id.md

```diff
@@ -0,0 +1,237 @@
+ID must have that Debug ID embedded to ensure the file is self-identifying.
+
+### Debug IDs in Source Maps
+
+We propose adding a `debugId` property to the source map at the top level of
+the source map object.  This property should be a string value representing
+the Debug ID in hexadecimal characters, preferably in the canonical UUID
+format:
+
+```json
+{
+  "version": 3,
+  "file": "app.min.js",
+  "debugId": "85314830-023f-4cf1-a267-535f4e37bb17",
+  "sources": [...],
+  "sourcesContent": [...],
+  "mappings": "..."
+}
+```
+
+### Debug IDs in JavaScript Artifacts
+
+Transpiled JavaScript files containing a Debug ID must embed the ID near the end
+of the source, ideally on the last line, in the format `//# debugId=<DEBUG_ID>`:
+
+```javascript
+//# debugId=85314830-023f-4cf1-a267-535f4e37bb17
+```
+
+If the special `//# sourceMappingURL=` comment already exists in the file, it is
+recommended to place the `debugId` comment in the line above to maintain
+compatibility with existing tools.  Because the last line already has meaning in
+the existing specification for the `sourceMappingURL` comment, tools are
+required to examine the last 5 lines to discover the Debug ID.
+
+## JavaScript API for Debug ID Resolution
+
+Today `error.stack` in most runtimes only returns the URLs of the files referenced
+by the stack trace.  For Debug IDs to be useful, a solution would need to be added
+to enable mapping of JavaScript file URLs to Debug IDs.
+
+The strawman proposal is to add the Debug ID in two locations:
+
+* `import.meta.debugId`: a new property that should return the debug ID as UUID
+  of the current module if has one
+* `System.getDebugIdForUrl(url)` looks up the debug ID for a given script file by
+  URL that has already been loaded by the browser in the current context.
+
+## Appendix A: Self-Description of Files
+
+Unfortunately, neither transpiled JavaScript files nor source maps can be easily
+identified without employing heuristics. Unlike formats like ELF binaries, they
+lack a distinctive header for identification purposes. When batch processing
+files, the ability to differentiate between various files is invaluable, but
+this capability is not fully realized in the context of source maps or
+transpiled JavaScript files. Although solving this issue is beyond the scope of
+this document, addressing it would significantly aid in distinguishing different
+files without relying on intricate heuristics.
+
+Nevertheless, we recommend that tools utilize the following heuristics to
+determine self-identifying JavaScript files and source maps:
+
+* a JSON file containing a toplevel object with the keys `mapping`, `version`,
+  `debugId` and `sourcesContent` should be considered to be a self-identifying
+  source map.
+* a UTF-8 encoded text file matching the regular expression
+  `(?m)^//# debugId=([a-fA-F0-9-]{12,})$` should be considered a transpiled
+  JavaScript file.
```

Comment:
👍🏻 I like this, especially also implicitly forcing JS files to be UTF-8 ;-)

Should we also propose a JSON Schema along with a `"$schema"` field for source maps, or would that be a bit too much?

Reply:
Mostly want to see initial feedback but a JSON schema for source maps would be very valuable by itself.